akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,365 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import java.net.InetSocketAddress
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import java.security.KeyStore
+import java.security.SecureRandom
+import javax.net.ssl._
+
+import scala.concurrent.Await
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration.Duration
+
+import akka.Done
+import akka.NotUsed
+import akka.actor.ExtendedActorSystem
+import akka.dispatch.ExecutionContexts
+import akka.event.Logging
+import akka.remote.RemoteActorRefProvider
+import akka.remote.RemoteTransportException
+import akka.remote.artery.Decoder.InboundCompressionAccess
+import akka.remote.artery.compress._
+import akka.stream.Attributes
+import akka.stream.Attributes.LogLevels
+import akka.stream.Client
+import akka.stream.IgnoreComplete
+import akka.stream.KillSwitches
+import akka.stream.Server
+import akka.stream.SharedKillSwitch
+import akka.stream.SinkShape
+import akka.stream.TLSProtocol._
+import akka.stream.TLSRole
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.Framing
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Keep
+import akka.stream.scaladsl.MergeHub
+import akka.stream.scaladsl.Partition
+import akka.stream.scaladsl.RestartFlow
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Source
+import akka.stream.scaladsl.TLS
+import akka.stream.scaladsl.Tcp
+import akka.stream.scaladsl.Tcp.ServerBinding
+import akka.util.ByteString
+
+/**
+ * INTERNAL API
+ */
+private[remote] class ArteryTcpTransport(_system: ExtendedActorSystem, _provider: RemoteActorRefProvider,
+                                         tlsEnabled: Boolean)
+  extends ArteryTransport(_system, _provider) {
+  import ArteryTransport.InboundStreamMatValues
+  import FlightRecorderEvents._
+
+  @volatile private var inboundConnectionsKillSwitch: SharedKillSwitch = KillSwitches.shared("inboundConnectionsKillSwitch")
+  private var serverBinding: Option[Future[ServerBinding]] = None
+
+  override protected def startTransport(): Unit = {
+    // FIXME new TCP events
+    topLevelFREvents.loFreq(Transport_AeronStarted, NoMetaData)
+  }
+
+  override protected def outboundTransportSink(
+    outboundContext: OutboundContext,
+    streamId:        Int,
+    bufferPool:      EnvelopeBufferPool): Sink[EnvelopeBuffer, Future[Done]] = {
+    implicit val sys = system
+
+    val host = outboundContext.remoteAddress.host.get
+    val port = outboundContext.remoteAddress.port.get
+
+    def tcp = Tcp()
+      .outgoingConnection(
+        remoteAddress = InetSocketAddress.createUnresolved(host, port),
+        halfClose = true, // issue https://github.com/akka/akka/issues/24392 if set to false
+        connectTimeout = Duration.Inf // FIXME should this be set, default is Inf?
```

Comment:
I can reproduce that with iptables. The HandshakeTimeoutException will take care of it, but I think it would be good to have a connectionTimeout. When that hits the RestartFlow will try again, and finally fail with HandshakeTimeoutException if still can't connect.